internal/dnfjson/cache.go

```diff
@@ -1,46 +1,194 @@
 package dnfjson
 
 import (
+	"fmt"
 	"io/fs"
 	"os"
 	"path/filepath"
+	"sort"
+	"time"
+
+	"github.com/gobwas/glob"
 )
 
+// A collection of directory paths, their total size, and their most recent
+// modification time.
+type pathInfo struct {
+	paths []string
+	size  uint64
+	mtime time.Time
+}
+
 type rpmCache struct {
 	// root path for the cache
 	root string
 
+	// individual repository cache data
+	repoElements map[string]pathInfo
+
+	// list of known repository IDs, sorted by mtime
+	repoRecency []string
+
+	// total cache size
+	size uint64
+
 	// max cache size
 	maxSize uint64
 }
 
 func newRPMCache(path string, maxSize uint64) *rpmCache {
-	return &rpmCache{
-		root:    path,
-		maxSize: maxSize,
+	r := &rpmCache{
+		root:         path,
+		repoElements: make(map[string]pathInfo),
+		size:         0,
+		maxSize:      maxSize,
 	}
+	// collect existing cache paths and timestamps
+	r.updateInfo()
+	return r
 }
 
-func (r *rpmCache) size() (uint64, error) {
-	var size uint64
-	sizer := func(path string, info fs.FileInfo, err error) error {
+func (r *rpmCache) clean() error {
+	curSize, err := dirSize(r.root)
+	if err != nil {
+		return err
+	}
+	if curSize > r.maxSize {
+		return os.RemoveAll(r.root)
+	}
+	return nil
+}
+
+// updateInfo updates the repoPaths and repoRecency fields of the rpmCache.
+func (r *rpmCache) updateInfo() {
+	cacheEntries, _ := os.ReadDir(r.root)
```

Comment:
Are any of the files that constitute a cache mandatory? Say the `.solv` ones? Could we simply do a glob over `*.solv` and then we would turn that into a list of repoid's by chopping off the `.solv` postfix? That way we would not have to care about the format/length of the repoid. We could then do the `updateInfo()` call per ID, where we would again glob over `<repoid>*`. Really a nitpick, this is fine.

Reply:
Oh yeah this is a good point.  I should read some manuals (or talk to the dnf devs).